Support batching using composite primary keys and multiple column ordering

TakuyaKurimoto opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->
Currently, when using the in_batches method on a model with composite primary keys (id_1, id_2), if :asc(:desc) is specified for the order argument, the values are returned with both id_1 and id_2 sorted in ascending (descending) order.
Allow selection of ascending or descending order for each key in composite primary key.

EXAMPLE
```
class Order < ActiveRecord::Base
  self.primary_key = [:id_1, :id_2]
end
Order.find_each(order: [:asc, :desc])
```
In the above code, id_1 is sorted in ascending order, id_2 in descending
### Detail

This Pull Request changes [REPLACE ME]

### Additional information

<!-- Provide additional information such as benchmarks, reference to other repositories or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

eileencodes commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -325,5 +351,12 @@ def act_on_ignored_order(error_on_ignore)
+        diff = Array(primary_key).size - orders.size
+        return orders if diff <= 0
+        orders + Array.new((diff), :asc)
+      end
```
This feels overly complex, if the application didn't explicitly set an order we can assume it's `asc` without having to build an array of the defaults. Can you find a way to simplify this when we want the default case?

nvasilevski replied:
It looks like there is a way for us to move the "default". handling to the `batch_order` method itself so it looks like

```ruby
Array(primary_key).zip(orders).map do |column, order| 
  o = order || DEFAULT_ORDER
  table[column].public_send(o)
end
```

We won't need much changes as we can rely on the current `zip` behavior as it fills missing values with `nil`:

```ruby
[:col1, :col2].zip([:asc]).to_h
{:col1=>:asc, :col2=>nil}
```

Though it means we'll need to find a similar approach to build the `operators` array. I'll try to leave a suggestion there as well

eileencodes reviewed (changes requested):
Hey @TakuyaKurimoto thanks for the PR and for working on this change. I left some comments about documentation improvements / style was well as one code comment.

nvasilevski commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -274,46 +292,54 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
         end
 
+        orders_size = orders.size
+        operators = orders.map.with_index(1) do |ord, index|
```
I generally prefer to avoid using indexes as they are very primitive and hide richer concepts. For example in our case indexes hide the fact that we want to treat "the last" object in a special way. It would be very helpful if we could write the code with an explicit "last concept"

Assuming that we won't be pre-building the `orders` array in `convert_order_param_to_array` anymore we could do it here like:
```ruby
orders = Array(primary_key).zip(orders)
last_order_value = orders.pop[1]
operators = orders.map do |column, order|
    o == :desc ? :lt : :gt
end
operators << last_order_value  == :desc ? :lteq : :gteq
```

We may need to extract `Array(primary_key).zip(orders)` and do the conversion once as it looks like we are using "orders by pk column" mapping in several places

Buyzed replied:
#47901

Buyzed replied:
(#47901)-

nvasilevski reviewed (commented):
The proposed API looks good to me. There is definitely an option to go with a hash-based API and require applications to explicitly specify which column the ordering value belongs to like `order: {col1: :asc, col2: :desc}` Hash-based API is more verbose but removes the need to know the order of the primary key. At this point an array based API seems to be perfectly sufficient so we shouldn't be overcomplicating it with a Hash-based version. Just wanted to mention it as a possible alternative or addition for the long-term

TakuyaKurimoto committed:
Support batching using composite primary keys and multiple column ordering

When find_each/find_in_batches/in_batches are performed on a table with composite primary keys, ascending or descending order can be selected for each key.

    ```ruby
    Person.find_each(order: [:desc, :asc]) do |person|
      person.party_all_night!
    end
    ```

TakuyaKurimoto commented:
@nvasilevski @eileencodes 
Thanks for the review! I learned a lot from it.
I have corrected the points you raised in your review, and I hope you will check it out when you have time.

nvasilevski commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -288,32 +324,39 @@ def apply_limits(relation, start, finish, order)
       end
 
       def apply_start_limit(relation, start, order)
```
Not critical but perhaps we could expect this method to receive a `Hash` of column=>order so we don't need to call `get_the_order_of_primary_key` in both methods. In this case `apply_limits` will have to change to something like:
```ruby
      def apply_limits(relation, start, finish, order)
        orders = get_the_order_of_primary_key(order) if start || finish
        relation = apply_start_limit(relation, start, orders) if start
        relation = apply_finish_limit(relation, finish, orders) if finish
        relation
      end
```

TakuyaKurimoto commented:
@nvasilevski @eileencodes
Thanks for the review despite your busy schedule!
I've made some modifications to the code and I hope you will check it out when you have time.

nvasilevski commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -274,46 +304,60 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
+        batch_order_array.map do |column, order|
+          operators[column] = order == :desc ? :lteq : :gteq
+        end
+        operators[last_column] = last_order == :desc ? :lt : :gt
```
Despite of ruby will enumerate over `Hash` in order of keys insertion I would prefer `operators` to stay as an array as on inspection it will be clearer which element exactly is the last while it may not be exactly clear if we use a hash.

Sorry, I might have said "hash" explicitly in my previous comments but ideally I think the data should be represented by an array

TakuyaKurimoto replied:
@nvasilevski
I did this because  [hash based api might be implemented in the future](https://github.com/rails/rails/pull/48268#pullrequestreview-1440030742), but I did not need to do it now, and you are right that an array is easy to understand the order. I fixed it, so please check it out.

nvasilevski commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -274,46 +304,58 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
-          relation, primary_key, primary_key_offset, order == :desc ? :lt : :gt
-        )
+        batch_orders_copy = batch_orders.dup
+        _last_column, last_order = batch_orders_copy.pop.first
```
I don't think this actually works as we don't need to call `.first` as long as we are splatting the array into two variables

```suggestion
        _last_column, last_order = batch_orders_copy.pop
```

Otherwise `last_order` will be `nil`:

```ruby
irb(main):008:0> batch_orders_copy = [[:col1, :asc], [:col2, :desc]]
=> [[:col1, :asc], [:col2, :desc]]
irb(main):009:0> _last_column, last_order = batch_orders_copy.pop.first
=> :col2
irb(main):010:0> _last_column
=> :col2
irb(main):011:0> last_order
=> nil
```

TakuyaKurimoto replied:
@nvasilevski
After [this comment](https://github.com/rails/rails/pull/48268#discussion_r1202822901) I changed the return value of the build_batch_orders method to the following
```
{:col1 => :asc, :col2 => :desc}
```
 and after [this comment](https://github.com/rails/rails/pull/48268#discussion_r1204636132) I changed the return value to the following. 
```
[{:col1 => :asc}, {:col2 => :desc}]
```
But maybe the return value you intended was something like the following?
```
[[:col1, :asc], [:col2, :desc]]
```
 If so, I will modify the code!

nvasilevski commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -274,46 +304,58 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
-        )
+        batch_orders_copy = batch_orders.dup
+        _last_column, last_order = batch_orders_copy.pop.first
+        operators = batch_orders_copy.map do |batch_order|
+          _column, order = batch_order.first
```
```suggestion
        operators = batch_orders_copy.map do |_column, order|
```

nvasilevski commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -274,46 +304,58 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
-      def apply_start_limit(relation, start, order)
-        batch_condition(relation, primary_key, start, order == :desc ? :lteq : :gteq)
+      def apply_start_limit(relation, start, batch_orders)
+        operators = batch_orders.map do |batch_order|
+          _column, order = batch_order.first
```
```suggestion
        operators = batch_orders.map do |_column, order |
```

nvasilevski commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -274,46 +304,58 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
-      def apply_finish_limit(relation, finish, order)
-        batch_condition(relation, primary_key, finish, order == :desc ? :gteq : :lteq)
+      def apply_finish_limit(relation, finish, batch_orders)
+        operators = batch_orders.map do |batch_order|
+          _column, order = batch_order.first
```
```suggestion
        operators = batch_orders.map do |_column, order|
```

TakuyaKurimoto commented:
@nvasilevski
After [this comment](https://github.com/rails/rails/pull/48268#discussion_r1202822901) I changed the return value of the build_batch_orders method to the following
```
{:col1 => :asc, :col2 => :desc}
```
 and after [this comment](https://github.com/rails/rails/pull/48268#discussion_r1204636132) I changed the return value to the following. 
```
[{:col1 => :asc}, {:col2 => :desc}]
```
But maybe the return value you intended was something like the following?
```
[[:col1, :asc], [:col2, :desc]]
```
 If so, I will modify the code!

nvasilevski commented:
@TakuyaKurimoto yeah, sorry I was under impression that we are going with `[[:col1, :asc], [:col2, :desc]]` structure as we accept `[:asc, :desc]` as an argument and after zipping it into the primary key we get `[[:col1, :asc], [:col2, :desc]]`

nvasilevski commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -274,46 +304,55 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
-        batch_relation = batch_condition(
-          relation, primary_key, primary_key_offset, order == :desc ? :lt : :gt
-        )
+        batch_orders_copy = batch_orders.dup
```
I was wondering if this is a precaution or we must `dup` the array as somewhere else we expect `batch_orders` to keep original size. 
I'm not concerned about it as I still find `.pop` solution being more readable compared to `current_index == last_index` so I'm okay with keeping `.dup` 

But I wouldn't be against getting back to index comparison approach if someone has strong opinion against `.dup`

nvasilevski reviewed (approved):
👍

TakuyaKurimoto commented:
@nvasilevski
thanks!
@eileencodes
I would be grateful if you would review the code.

eileencodes reviewed (approved):
Thank you @TakuyaKurimoto!
